include/sisl/fds/concurrent_insert_vector.hpp

```diff
@@ -0,0 +1,122 @@
+/*********************************************************************************
+ * Modifications Copyright 2017-2019 eBay Inc.
+ *
+ * Author/Developer(s): Harihara Kadayam
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *    https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed
+ * under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+ * CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ *
+ *********************************************************************************/
+#pragma once
+
+#include <cstdint>
+#include <vector>
+
+#include <sisl/utility/thread_buffer.hpp>
+
+namespace sisl {
+
+//
+// This data structure provides a vector where concurrent threads can safely emplace or push back the data into.
+// However, it does not guarantee any access or iterations happen during the insertion. It is the responsibility of the
+// user to synchornoize this behavior. This data structure is useful when the user wants to insert data into a vector
+// concurrently in a fast manner and then iterate over the data later. If the user wants a vector implementation which
+// reads concurrently with writer, they can use sisl::ThreadVector. This data structure is provided as a replacement for
+// simplistic cases where insertion and iteration never happen concurrently. As a result it provides better performance
+// than even sisl::ThreadVector and better debuggability.
+//
+// Benchmark shows atleast 10x better performance on more than 4 threads concurrently inserting with mutex.
+//
+template < typename T >
+class ConcurrentInsertVector {
+private:
+    ExitSafeThreadBuffer< std::vector< T >, size_t > tvector_;
+    std::vector< std::vector< T > const* > per_thread_vec_ptrs_;
+
+public:
+    struct iterator {
+        size_t next_thread{0};
+        size_t next_id_in_thread{0};
+        ConcurrentInsertVector const& vec;
+
+        iterator(ConcurrentInsertVector const& v) : vec{v} {}
+        iterator(ConcurrentInsertVector const& v, bool end_iterator) : vec{v} {
+            if (end_iterator) { next_thread = vec.per_thread_vec_ptrs_.size(); }
+        }
+
+        void operator++() {
+            ++next_id_in_thread;
+            if (next_id_in_thread >= vec.per_thread_vec_ptrs_[next_thread]->size()) {
+                ++next_thread;
+                next_id_in_thread = 0;
+            }
+        }
+
+        bool operator==(iterator const& other) const {
+            return (next_thread == other.next_thread) && (next_id_in_thread == other.next_id_in_thread);
```

Comment:
Is not comparing the `vec` field part of this wizardry? If not, can we just use synthesized equality operators?

Reply:
If I understand your comment correctly, you are asking how do we handle cases where iterators from 2 different ConcurrentInsertVector are compared? If so yes, that is a very good catch and we need to protect ourselves on that. Let me add that.

Reply:
Yes, that was the first part; and then we seem to compare each field in declaration order anyways so what is wrong with `default`?

Reply:
Yes you are right, I started with non-specialized constructor for end() iterator, but now that we have one, == override isn't needed.